### ISSUE TITLE
Bogus void_checks warning when a FutureOr<void> callback contains a throw

### ISSUE THREAD

**Issue by nex3** [dart-lang/linter#3172]
The following code:

```dart
void capture(FutureOr<void> Function() callback) {}

void main() {
  capture(() {
   throw "oh no";
  });
}
```

produces this analyzer output:

```
Analyzing test.dart...

   info - test.dart:128:11 - Don't assign to void. - void_checks

1 issue found.
```

This is clearly incorrect; the callback returns `void`, which is a valid subtype of `FutureOr<void>`. Removing the `throw` or making the callback `async` makes the warning disappear.

**Comment by eernstg** [comment]
The return type of the function literal `() { throw "oh no"; }` would be `Never` rather than `void`. 

The reason for this is that the function literal return type is [computed](https://github.com/dart-lang/language/blob/master/resources/type-system/inference.md#function-literal-return-type-inference) based on the 'actual returned type' (as well as a couple of other things). The actual returned type starts off as `Never` in this case (because we can't reach the end of the function), and it gets updated to a supertype with each return statement. We have no return statements here, so the actual returned type is `Never`.

Then we have the following rule:

> if `R` is `void`, or the function literal is marked `async` and `R` is `FutureOr<void>`, let `S` be `void`.

where `R` is the return type taken from the greatest closure of the context type of the function literal. In this case `R` is `FutureOr<void>`, but the function literal is not `async`, so the rule does not apply, and we continue:

> Otherwise, if `T <: R` then let `S` be `T`. 
> ...
> [For the given function literal] the inferred return type is `S`.

`T` is the actual returned type, and `Never <: R` is true for all `R`, so the inferred return type is `Never`.

We did adjust these rules a couple of times, and it is possible that we could get better results by making yet another small adjustment. But the first thing to note would be that it seems to be working as intended. If we do update the rules then we should be ready to handle the breakage (this kind of change tends to make things slightly different in many locations, so it's not obvious that we'd want to accept the breakage unless it turns out to be a significant improvement).

**Comment by nex3** [comment]
It makes sense that the inferred return type is `Never`, but why is it a problem to assign `Never` to `FutureOr<void>`? That seems like the heart of the issue here.

**Comment by eernstg** [comment]
Ah, I only looked at the part that seemed to be connected to the language specification, and I forgot the part that's directly about the lint. With respect to the lint, of course, I'd just be commenting on it as an outsider&mdash;I'm not working on the linter, and I don't know exactly what this particular lint is intended to do. Anyway, it is tempting to add a couple of comments:

I can't actually see how the example could be considered to assign anything to void, so I agree that it looks like there is an issue with `void_checks`. In particular, we don't get the lint on the following variant:

```dart
import 'dart:async';

void capture(FutureOr<void> Function() callback) {}
Never f() { throw 0; } // Same type as the function literal.

void main() {
  capture(f); // No lint.
}
```

Checking the behavior of `void_checks`, it does _not_ flag higher-order cases (e.g., assigning an expression of type `int Function(int)` to a variable of type `void Function(int)`, or, for the contravariant case, an expression of type `int Function(void)` to a variable of type `int Function(int)`):

```dart
void f(List<void> xs) {}

void main() {
  f(<int>[]); // No lint.
}
```

So we'd need to find an assignment of an expression of a type which isn't void to a variable/parameter/return whose declared type is `void`, and I don't see this anywhere in the original example.

As for the treatment of `Never`:

> why is it a problem to assign `Never` to `FutureOr<void>`?

I don't think it is a problem, and we don't get any diagnostic messages from the following:

```dart
import 'dart:async';

void main() {
  late FutureOr<void> x = throw 0;
  print(x);
}
```
